### ISSUE TITLE
dart2js: target modern browsers and use modern JS

### ISSUE THREAD

**Issue by sigmundch** [dart-lang/sdk#45092]
[Two years ago][1] we announced a plan to deprecate support for IE11, but due to change in priorities we have not yet made the underlying changes.

This bug is to track any work to enable modern JS features and start using them in our compiler. This includes:
* [ ] increase test coverage of `--no-legacy-javascript`
* [ ] fix current failures with the flag
* [ ] explore other features we'd like to include in an initial version of this flag, this may include:
  * [ ] syntax changes (e.g. arrow function has come up, but it requires careful integration with internal stack deobfuscation tools)
  * [ ] use of symbols for various reasons (e.g. today we feature detect to use it in rti)
  * [ ] Use of ES6 maps
  * [ ] Removing IE11 polyfills in dart:html
* [ ] measure performance and code-size changes
* [ ] dogfood the flag on a couple internal apps
* [ ] enable the flag by default

[1]: https://groups.google.com/a/dartlang.org/g/announce/c/x7eDinVT6fM/m/ZSFl2a9tEAAJ

**Comment by lexaknyazev** [comment]
Please explore including #29623 in this plan.

**Comment by sigmundch** [comment]
Thanks for the suggestion, I went ahead and added it to the list above.

**Comment by MitaiGit** [comment]
it is very sad that you have to adapt to a poor javascript system, instead of developing your own virtual machine for the browser

**Comment by lexaknyazev** [comment]
Please consider promoting BigInt into language. With modern JS, it implies:
- Literals and const values: #33893
- JS-interop (to avoid [workarounds](https://github.com/dart-lang/sdk/issues/32399#issuecomment-828790308))
- 64-bit typed data via `BigInt64Array` / `BigUint64Array`

**Comment by MrBirb** [comment]
I believe in 2.14 this is the default behavior?

**Comment by sigmundch** [comment]
Yes, the flag was flipped by default in 2.14. So far, though, we only use ES6 method syntax, which was the change that was providing the biggest win in terms of code-size reduction (as much as 4% on some of our large internal apps).

**Comment by DynamicProdBreaker** [comment]
Would ES2015 modules be of any use to Dart? I imagine it would help with global variables, generated code would be able to be split up over several files if necessary, and Dart wouldn't have to wrap the entire source in a function expression to be executed, at the very least.
